### PR TITLE
Bump compiler and analyzer versions

### PR DESCRIPTION
**Pull request by drewnoakes**
The newer compiler and analyzers produced new diagnostics on our code, which have also been addressed to remain free of warnings.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/project-system/pull/8598)

### CODE REVIEW

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ExternalContracts.cs [hunk 1]
```diff
@@ -9,6 +9,8 @@
 using NuGet.SolutionRestoreManager;
 using NuGet.VisualStudio;
 
+#pragma warning disable RS0030 // Do not used banned APIs
```
Comment (tmeschter) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ExternalContracts.cs:12]: Which banned API was it complaining about?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/WPF/Converters/Converters.LambdaConverter.cs [hunk 2]
```diff
@@ -15,15 +15,15 @@ internal static partial class Converters
...
+            private readonly bool _convertFromNull;
 
             public LambdaConverter(Func<TFrom, TTo> convert, Func<TTo, TFrom>? convertBack = null, bool convertFromNull = true)
             {
-                this.convert = convert;
-                this.convertBack = convertBack;
-                this.convertFromNull = convertFromNull;
+                this._convert = convert;
+                this._convertBack = convertBack;
+                this._convertFromNull = convertFromNull;
```
Comment (tmeschter) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/WPF/Converters/Converters.LambdaConverter.cs:26]: Nit: still has `this`.

**Review by tmeschter — APPROVED** [review]

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ExternalContracts.cs (hunk 1 above)
Comment (drewnoakes) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ExternalContracts.cs:12]: Three different ones. I didn't think it was useful to highlight them in this file.

Comment (tmeschter) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ExternalContracts.cs:12]: Should there be a `#pragma warning restore RS0030` somewhere? I assume it doesn't really matter in this file.

Comment (drewnoakes) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ExternalContracts.cs:12]: I don't think it's necessary in this file. So long as we only ever model external contracts here.

This file's just adding metadata to external contracts, which is still useful, even if we have banned their use.

Most of these banned contracts are actually used indirectly, where we wrap them in an unbanned API that gives some guarantee we want to have, and suppress RS0030 on that usage.

**Pull request merged**